File: demo/src/main/java/com/bytedance/scenedemo/CaseListDemo.java
package com.bytedance.scenedemo;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ScrollView;

import com.bytedance.scene.group.UserVisibleHintGroupScene;
import com.bytedance.scenedemo.case0.Case0Scene;
import com.bytedance.scenedemo.case0.Case1Scene;
import com.bytedance.scenedemo.case0.Case2Scene;
import com.bytedance.scenedemo.case0.Case3Scene;
import com.bytedance.scenedemo.case0.Case4Scene;
import com.bytedance.scenedemo.case0.Case5Scene;

/**
 * Created by JiangQi on 8/21/18.
 */
public class CaseListDemo extends UserVisibleHintGroupScene {
    @NonNull
    @Override
    public ViewGroup onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        ScrollView scrollView = new ScrollView(getActivity());

        LinearLayout layout = new LinearLayout(getActivity());
        layout.setOrientation(LinearLayout.VERTICAL);

        scrollView.addView(layout);

        Button button = new Button(getActivity());
        button = new Button(getActivity());
        button.setAllCaps(false);
        button.setText("Push后立刻Pop，循环100次，动画要正确");
        layout.addView(button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getNavigationScene().push(Case0Scene.class);
            }
        });

        button = new Button(getActivity());
        button.setAllCaps(false);
        button.setText("同时Push 100个");
        layout.addView(button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getNavigationScene().push(Case1Scene.class);
            }
        });

        button = new Button(getActivity());
        button.setAllCaps(false);
        button.setText("同时多个Pop超过当前的Scene数量");
        layout.addView(button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getNavigationScene().push(Case2Scene.class);
            }
        });

        button = new Button(getActivity());
        button.setAllCaps(false);
        button.setText("Push和Pop动画过程中，Remove");
        layout.addView(button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getNavigationScene().push(Case3Scene.class);
            }
        });

        button = new Button(getActivity());
        button.setAllCaps(false);
        button.setText("Scene 生命周期内又执行Push/Pop");
        layout.addView(button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getNavigationScene().push(Case4Scene.class);
            }
        });

        button = new Button(getActivity());
        button.setAllCaps(false);
        button.setText("GroupScene 没添加到Scene之前就add/remove");
        layout.addView(button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getNavigationScene().push(Case5Scene.class);
            }
        });

        button = new Button(getActivity());
        button.setAllCaps(false);
        button.setText("Activity finish后立刻 Push/Pop");
        layout.addView(button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getActivity().finish();
                getNavigationScene().push(Case0Scene.EmptyScene.class);
            }
        });

        return scrollView;
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (isVisible() && isVisibleToUser) {

        }
    }
}

File: demo/src/main/java/com/bytedance/scenedemo/CaseListDemo.kt
package com.bytedance.scenedemo

import com.bytedance.scene.group.UserVisibleHintGroupScene
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import android.content.Intent
import android.view.View
import android.widget.*
import com.bytedance.scene.ktx.requireNavigationScene
import com.bytedance.scenedemo.auto_recycle.AutoRecycleActivity
import com.bytedance.scenedemo.restore.SupportRestoreActivity
import com.bytedance.scenedemo.extreme_case.Case0Scene
import com.bytedance.scenedemo.extreme_case.Case1Scene
import com.bytedance.scenedemo.extreme_case.Case2Scene
import com.bytedance.scenedemo.extreme_case.Case3Scene
import com.bytedance.scenedemo.extreme_case.Case4Scene
import com.bytedance.scenedemo.extreme_case.Case5Scene
import com.bytedance.scenedemo.utility.addClassPathTitle

/**
 * Created by JiangQi on 8/21/18.
 */
class CaseListDemo : UserVisibleHintGroupScene() {
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup, savedInstanceState: Bundle?): ViewGroup {
        val scrollView = ScrollView(activity)
        val layout = LinearLayout(activity)
        layout.orientation = LinearLayout.VERTICAL
        scrollView.addView(layout)
        addClassPathTitle(layout)
        addSpace(layout, 12)
        addTitle(layout, getString(R.string.main_title_case))
        addButton(
            layout,
            getString(R.string.main_case_destroy_restore),
            View.OnClickListener {
                requireNavigationScene().startActivity(
                    Intent(
                        requireActivity(),
                        SupportRestoreActivity::class.java
                    )
                )
            })
        addButton(
            layout,
            getString(R.string.main_case_push_and_pop),
            View.OnClickListener { requireNavigationScene().push(Case0Scene::class.java) })
        addButton(
            layout,
            getString(R.string.main_case_push_many),
            View.OnClickListener { requireNavigationScene().push(Case1Scene::class.java) })
        addButton(
            layout,
            getString(R.string.main_case_pop_many),
            View.OnClickListener { requireNavigationScene().push(Case2Scene::class.java) })
        addButton(
            layout,
            getString(R.string.main_case_push_pop_remove),
            View.OnClickListener { requireNavigationScene().push(Case3Scene::class.java) })
        addButton(
            layout,
            getString(R.string.main_case_push_pop_in_lifecycle),
            View.OnClickListener { requireNavigationScene().push(Case4Scene::class.java) })
        addButton(
            layout,
            getString(R.string.main_case_add_remove_group_scene),
            View.OnClickListener { requireNavigationScene().push(Case5Scene::class.java) })
        addButton(layout, getString(R.string.main_case_push_pop_after_finish), View.OnClickListener {
            requireActivity().finish()
            requireNavigationScene().push(Case0Scene.EmptyScene::class.java)
        })
        addButton(
            layout,
            getString(R.string.main_case_auto_recycle_scene_when_low_memory),
            View.OnClickListener {
                requireNavigationScene().startActivity(
                    Intent(
                        requireActivity(),
                        AutoRecycleActivity::class.java
                    )
                )
            }
        )
        addSpace(layout, 100)
        return scrollView
    }

    private fun addTitle(parent: LinearLayout, text: String) {
        val textView = TextView(activity)
        textView.textSize = 14f
        textView.text = text
        val lp = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        lp.leftMargin = 30
        lp.rightMargin = 30
        lp.topMargin = 24
        lp.bottomMargin = 24
        parent.addView(textView, lp)
    }

    private fun addButton(parent: LinearLayout, text: String, onClickListener: View.OnClickListener): Button {
        val button = Button(activity)
        button.isAllCaps = false
        button.text = text
        button.setOnClickListener(onClickListener)
        val lp = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 150)
        lp.leftMargin = 20
        lp.rightMargin = 20
        parent.addView(button, lp)
        return button
    }

    private fun addSpace(parent: LinearLayout, height: Int) {
        val space = Space(activity)
        parent.addView(space, ViewGroup.LayoutParams.MATCH_PARENT, height)
    }
}